valhalla/meili/routing.h

```diff
@@ -23,107 +23,139 @@ namespace meili {
-        predecessor(the_predecessor),
-        edgelabel(the_edgelabel) {
-    if (!((nodeid.Is_Valid() && dest == kInvalidDestination) || (!nodeid.Is_Valid() && dest != kInvalidDestination))) {
-      throw std::invalid_argument("nodeid and dest must be mutually exclusive, i.e. either nodeid is valid or dest is valid");
-    }
-
+  Label(const baldr::GraphId& nodeid, uint16_t dest,
+        const baldr::GraphId& edgeid, float source, float target,
+        const sif::Cost& cost, float turn_cost, float sortcost,
+        const uint32_t predecessor, const baldr::DirectedEdge* edge,
+        const sif::TravelMode mode)
+     : sif::EdgeLabel(predecessor, edgeid, edge, cost, sortcost, 0.0f, mode, 0),
+       nodeid_(nodeid),
+       dest_(dest),
+       source_(source),
+       target_(target),
+       turn_cost_(turn_cost) {
+    // Validate inputs
     if (!(0.f <= source && source <= target && target <= 1.f)) {
-      throw std::invalid_argument("invalid source ("
-                                  + std::to_string(source)
-                                  + ") or target ("
-                                  + std::to_string(target) + ")");
+      throw std::invalid_argument("invalid source (" + std::to_string(source)
+                   + ") or target (" + std::to_string(target) + ")");
     }
-
     if (cost.cost < 0.f) {
       throw std::invalid_argument("invalid cost = " + std::to_string(cost.cost));
     }
-
     if (turn_cost < 0.f) {
       throw std::invalid_argument("invalid turn_cost = " + std::to_string(turn_cost));
     }
+  }
 
-    if (!edgelabel && the_edge) {
-      // Populate the sif EdgeLabel. The cost within Cost is the distance.
-      edgelabel = std::make_shared<const sif::EdgeLabel>(the_predecessor,
-           the_edgeid, the_edge, the_cost, sortcost, 0.0f, the_travelmode, 0);
-    }
+  /**
+   * Set the node Id.
+   * @param  id   Node Id.
+   */
+  void set_nodeid(const baldr::GraphId& id) {
+    nodeid_= id;
   }
 
-  // Must be mutually exclusive, i.e. nodeid.Is_Valid() XOR dest != kInvalidDestination
-  baldr::GraphId nodeid;
-  uint16_t dest;
+  /**
+   * Get the node Id.
+   * @return  Returns the node Id.
+   */
+  baldr::GraphId nodeid() const {
+    return nodeid_;
+  }
 
-  // Invalid graphid if dummy
-  baldr::GraphId edgeid;
+  /**
+   * Get the edge Id. If the predecessor is invalid return an invalid edgeid.
+   * This is used to handle internal map matching logic that relies on the
+   * first edge having an invalid edge id, but internal to routing the edgeid
+   * should be valid if passed in from a prior route state (for costing).
+   */
+  baldr::GraphId label_edgeid() const {
+    return predecessor_ == baldr::kInvalidLabel ?  baldr::GraphId() : edgeid();
+  }
 
-  // Assert: 0.f <= source <= target <= 1.f
-  float source;
-  float target;
+  /**
+   * Set the destination index.
+   * @param  dest  Destination index.
+   */
+  void set_dest(const uint16_t dest) {
```

Comment:
This setter can be removed too if the corresponding changes are made. Keep it immutable!